demos/sample_azure_iot/sample_azure_iot.c

```diff
@@ -0,0 +1,625 @@
+         * until  the maximum number of attempts are reached or the maximum timeout
+         * value is reached. The function returns a failure status if the TCP
+         * connection cannot be established to the broker after the configured
+         * number of attempts. */
+        xNetworkStatus = prvConnectToServerWithBackoffRetries( ( const char * ) pucIotHubHostname,
+                                                               democonfigMQTT_BROKER_PORT,
+                                                               &xNetworkCredentials, &xNetworkContext );
+        configASSERT( xNetworkStatus == TLS_TRANSPORT_SUCCESS );
+
+        /* Fill in Transport Interface send and receive function pointers. */
+        xTransport.pxNetworkContext = &xNetworkContext;
+        xTransport.xSend = TLS_Socket_Send;
+        xTransport.xRecv = TLS_Socket_Recv;
+
+        /* Init IoT Hub option */
+        xResult = AzureIoTHubClient_OptionsInit( &xHubOptions );
+        configASSERT( xResult == eAzureIoTHubClientSuccess );
+
+        xHubOptions.pucModuleID = ( const uint8_t * ) democonfigMODULE_ID;
+        xHubOptions.ulModuleIDLength = sizeof( democonfigMODULE_ID ) - 1;
+
+        xResult = AzureIoTHubClient_Init( &xAzureIoTHubClient,
+                                          pucIotHubHostname, pulIothubHostnameLength,
+                                          pucIotHubDeviceId, pulIothubDeviceIdLength,
+                                          &xHubOptions,
+                                          ucSharedBuffer, sizeof( ucSharedBuffer ),
+                                          ullGetUnixTime,
+                                          &xTransport );
+        configASSERT( xResult == eAzureIoTHubClientSuccess );
+
+        #ifdef democonfigDEVICE_SYMMETRIC_KEY
+            xResult = AzureIoTHubClient_SetSymmetricKey( &xAzureIoTHubClient,
+                                                         ( const uint8_t * ) democonfigDEVICE_SYMMETRIC_KEY,
+                                                         sizeof( democonfigDEVICE_SYMMETRIC_KEY ) - 1,
+                                                         Crypto_HMAC );
+            configASSERT( xResult == eAzureIoTHubClientSuccess );
+        #endif /* democonfigDEVICE_SYMMETRIC_KEY */
+
+        /* Sends an MQTT Connect packet over the already established TLS connection,
+         * and waits for connection acknowledgment (CONNACK) packet. */
+        LogInfo( ( "Creating an MQTT connection to %s.\r\n", pucIotHubHostname ) );
+
+        xResult = AzureIoTHubClient_Connect( &xAzureIoTHubClient,
+                                             false, &xSessionPresent,
+                                             sampleazureiotCONNACK_RECV_TIMEOUT_MS );
+        configASSERT( xResult == eAzureIoTHubClientSuccess );
+
+        xResult = AzureIoTHubClient_SubscribeCloudToDeviceMessage( &xAzureIoTHubClient, prvHandleCloudMessage,
```

Comment:
Do we really want to block the call for `sampleazureiotWAIT_FOREVER`?  I'm not sure that's great practice since "stuff happens".  We don't block forever on CONNAC, its configurable, i think there could be one sampleblahBlah_SUBSCRIBE_RECV_TIMEOUT for all the subscribers (don't need a per function timeout i don't tihkn).